### PR TITLE
Add early termination for `NeverType` match expression

### PR DESCRIPTION
**Pull request by rajyan**
fixes https://github.com/phpstan/phpstan/issues/6257
fixes https://github.com/phpstan/phpstan/issues/6251

### CODE REVIEW

File: src/Analyser/NodeScopeResolver.php [hunk 1]
```diff
@@ -1664,6 +1664,9 @@ private function findEarlyTerminatingExpr(Expr $expr, Scope $scope): ?Expr
 		}
 
 		$exprType = $scope->getType($expr);
+		if ($expr instanceof Expr\Match_ && $exprType instanceof NeverType) {
+			return $expr;
```
Comment (ondrejmirtes) [inline src/Analyser/NodeScopeResolver.php:1668]: I'm not sure about this fix. What should be done instead is that these situations should be caught by the `if ($exprType instanceof NeverType && $exprType->isExplicit()) {` instead.

"Implicit" NeverType is for example when you try to intersect two incompatible types (`string&int`). "Explicit" mixed is when you have a `throw` expression, function with `@return never` and similar.

This means that the explicitness of "NeverType" gets lost when we have a match expression with "explicit never" arms, and that's something that should be fixed instead.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/Analyser/NodeScopeResolver.php (hunk 1 above)
Comment (rajyan) [reply src/Analyser/NodeScopeResolver.php]: Thank you for the review and informative examples. I'll try fixing in that way!

**Comment by rajyan** [comment]
The cause is here, 
https://github.com/phpstan/phpstan-src/blob/51e2df38ee7fbda5416e1c5c1091dacf8618f3e1/src/Type/TypeCombinator.php#L134-L137
https://github.com/phpstan/phpstan-src/blob/51e2df38ee7fbda5416e1c5c1091dacf8618f3e1/src/Type/TypeCombinator.php#L304
but not sure how to fix yet.

Also, the comment says that it's transforming `A | never to A` https://github.com/phpstan/phpstan-src/blob/51e2df38ee7fbda5416e1c5c1091dacf8618f3e1/src/Type/TypeCombinator.php#L272
but `NeverType` is not handled there.

I think it can be handled like the comment says, and `never | never` can be `NeverType` with `$type->isExplicit() && $type->isExplicit()`.

But if I fix like that, test fails here with `array{Never, Never}` https://github.com/phpstan/phpstan-src/blob/86fd620d95a13ee9e8c38d5f734f0975e6074ad1/tests/PHPStan/Analyser/data/bug-1516.php#L1-L37.

**Comment by rajyan** [comment]
Have to take more time breaking and fixing `TypeCombinator` to understand it deeper...

**Comment by ondrejmirtes** [comment]
Hi, I've took the liberty to push a change here. Some tests still fail but I like it beter. Before TypeCombinator removed NeverType from `$types` so the fact the it's an explicit NeverType got lost.

Right now we leave NeverType in there so if `TypeCombinator::union()` is called with multiple explicit NeverType instances, the explicitness is preserved.

But some tests currently fail and I don't know why. Please debug that, thanks :)

Please read more about what's going on in TypeCombinator here: https://phpstan.org/developing-extensions/type-system (especially https://phpstan.org/developing-extensions/type-system#type-normalization).

**Comment by rajyan** [comment]
@ondrejmirtes 
Hi, thank you for looking into this PR.
The failing test were from `NeverType` handling, which `TypeCombinator` says that
https://github.com/phpstan/phpstan-src/blob/c4ac24bb834ad11bebc3baa5bd695afb718cd138/src/Type/TypeCombinator.php#L268
but not transformed actually.

By adding `NeverType` union handling https://github.com/phpstan/phpstan-src/pull/1078/commits/c4ac24bb834ad11bebc3baa5bd695afb718cd138, all tests passed except this one.
https://github.com/phpstan/phpstan-src/blob/86fd620d95a13ee9e8c38d5f734f0975e6074ad1/tests/PHPStan/Analyser/data/bug-1516.php#L1-L37

This one failed because it has a different reason related to `array_merge` (the test is green if `array_merge` is removed).
Empty array for `ConstantArray` was changed to and explicit `NeverType` in this commit https://github.com/phpstan/phpstan-src/commit/c2b9e713daa09db18111d28ba5c8cd79beeaeec5

but the handling for `ArrayMergeFunctionDynamicReturnTypeExtension`
https://github.com/phpstan/phpstan-src/blob/f2410518e8b8e558c5fd64fd04560c22c19bbad6/src/Type/Php/ArrayMergeFunctionDynamicReturnTypeExtension.php#L58-L60
is expecting a implicit `NeverType`

I hope this fix is in the right direction. I'll add a regression test for `NeverType` (and maybe for `ArrayMergeFunctionDynamicReturnTypeExtension` ?) if it looks good to you!

**Comment by ondrejmirtes** [comment]
The thing is - NeverType is already handled by `TypeCombinator::compareTypesInUnion()` by these lines:

```php
		if ($b->isSuperTypeOf($a)->yes()) {
			return [null, $b];
		}

		if ($a->isSuperTypeOf($b)->yes()) {
			return [$a, null];
		}
```

The root issue was that some types wrongly responded to isSuperTypeOf with NeverType. So I addressed the root issue and merged it: https://github.com/phpstan/phpstan-src/commit/99951ac47892e6f1896e3233cffaac9317b8daf0

Thank you.

**Comment by rajyan** [comment]
Thank you.
Sorry I couldn't get to the root issue...

**Pull request closed without merging**